packages/datastore/src/datastore/datastore.ts

```diff
@@ -620,6 +637,50 @@ const createModelClass = <T extends PersistentModel>(
 	return clazz;
 };
 
+export class AsyncCollection<T> implements AsyncIterable<T> {
+	start: number;
+	end: number;
+	values: Array<any>;
+	constructor(values: Array<any>) {
+		this.start = 0;
+		this.end = values.length;
+		this.values = values;
+	}
+	[Symbol.asyncIterator](): AsyncIterator<T> {
+		let index = this.start;
+		return {
+			next: async () => {
+				if (index < this.end) {
+					const result = {
+						value: this.values[index],
+						done: false,
+					};
+					index++;
+					return result;
+				}
+				return {
+					value: null,
+					done: true,
+				};
+			},
+		};
+	}
+	async toArray(arg?): Promise<T[]> {
+		const max = arg && arg.hasOwnProperty('max') ? arg.max : this.values.length;
```

Comment:
Let's make this typed, please:

```suggestion
	async toArray({max = Number.MAX_SAFE_INTEGER}: {max?: number} = {}): Promise<T[]> {
```

Reply:
Got it!